# Inventory Analysis

The purpose of this notebook is to understand the inventory data and also clean and prep the data.


In [4]:
import pandas as pd
import numpy as np

# Load the data
print("🔍 Loading current_inventory.csv...")
df = pd.read_csv('current_inventory.csv')

# Basic info
print(f"✅ Data loaded successfully!")
print(f"📊 Shape: {df.shape} (rows, columns)")
print(f"📋 Columns: {list(df.columns)}")

# Show first few rows
print("\n📋 First 3 rows:")
print(df.head(3))

🔍 Loading current_inventory.csv...
✅ Data loaded successfully!
📊 Shape: (1012, 19) (rows, columns)
📋 Columns: ['1009', '1009.1', '6655', 'Unnamed: 3', '5745', '0', '-9', '-1', '-4', '0.1', '-9.1', '-2', '0.2', '0.3', '-4.1', '0.4', '0.5', '0.6', '0.7']

📋 First 3 rows:
             1009                          1009.1           6655  \
0    Product Name                            msku  Opening Stock   
1  Chimmy Pillow   CSTE_0001_ST_Bts_Pillow_Chimmy            192   
2    Tata Pillow     CSTE_0002_ST_Bts_Pillow_Tata              0   

     Unnamed: 3  5745     0    -9    -1    -4   0.1  -9.1    -2   0.2   0.3  \
0  Buffer Stock  TLCQ  BLR7  BLR8  BOM5  BOM7  CCU1  CCX1  DEL4  DEL5  DEX3   
1           NaN   192     0     0     0    -1     0     0     0     0     0   
2           NaN     0     0     0     0     0     0     0     0     0     0   

   -4.1   0.4   0.5   0.6   0.7  
0  PNQ2  PNQ3  SDED  SDEE  XHJ9  
1     0     0     0     0     0  
2     0     0     0     0     0  


In [5]:
# Fix the header issue - use the first row as column names
print("🔧 Fixing column headers...")

# Get the actual column names from the first row
actual_columns = df.iloc[0].tolist()
print(f"📋 Actual columns found: {actual_columns}")

# Set these as the column names and drop the header row
df.columns = actual_columns
df = df.drop(0).reset_index(drop=True)

print(f"\n✅ Headers fixed!")
print(f"📊 New shape: {df.shape}")
print(f"📋 Corrected columns: {list(df.columns)}")

# Show first few rows with correct headers
print("\n📋 First 3 rows with correct headers:")
print(df.head(3))

🔧 Fixing column headers...
📋 Actual columns found: ['Product Name', 'msku', 'Opening Stock', 'Buffer Stock', 'TLCQ', 'BLR7', 'BLR8', 'BOM5', 'BOM7', 'CCU1', 'CCX1', 'DEL4', 'DEL5', 'DEX3', 'PNQ2', 'PNQ3', 'SDED', 'SDEE', 'XHJ9']

✅ Headers fixed!
📊 New shape: (1011, 19)
📋 Corrected columns: ['Product Name', 'msku', 'Opening Stock', 'Buffer Stock', 'TLCQ', 'BLR7', 'BLR8', 'BOM5', 'BOM7', 'CCU1', 'CCX1', 'DEL4', 'DEL5', 'DEX3', 'PNQ2', 'PNQ3', 'SDED', 'SDEE', 'XHJ9']

📋 First 3 rows with correct headers:
     Product Name                            msku Opening Stock Buffer Stock  \
0  Chimmy Pillow   CSTE_0001_ST_Bts_Pillow_Chimmy           192          NaN   
1    Tata Pillow     CSTE_0002_ST_Bts_Pillow_Tata             0          NaN   
2    Koya Pillow     CSTE_0003_ST_Bts_Pillow_Koya             0          NaN   

  TLCQ BLR7 BLR8 BOM5 BOM7 CCU1 CCX1 DEL4 DEL5 DEX3 PNQ2 PNQ3 SDED SDEE XHJ9  
0  192    0    0    0   -1    0    0    0    0    0    0    0    0    0    0  
1    0    0  

### Discovery

I count 15 warehouse locations across Indian cities:

1. BLR = Bangalore (BLR7, BLR8)
2. BOM = Mumbai/Bombay (BOM5, BOM7)
3. CCU/CCX = Kolkata/Calcutta (CCU1, CCX1)
4. DEL/DEX = Delhi (DEL4, DEL5, DEX3)
5. PNQ = Pune (PNQ2, PNQ3)
6. TLCQ, SDED, SDEE, XHJ9 = Other cities

Now let's analyze the data types and stock distribution:

In [6]:
# Check data types and convert warehouse columns to numeric
print("🔍 Checking data types...")
print(df.dtypes)

print("\n🔢 Converting warehouse columns to numeric...")

# List all warehouse columns (everything except Product Name, msku, Opening Stock, Buffer Stock)
warehouse_cols = ['TLCQ', 'BLR7', 'BLR8', 'BOM5', 'BOM7', 'CCU1', 'CCX1', 
                  'DEL4', 'DEL5', 'DEX3', 'PNQ2', 'PNQ3', 'SDED', 'SDEE', 'XHJ9']

# Convert to numeric, replacing any non-numeric values with 0
for col in warehouse_cols:
    df[col] = pd.to_numeric(df[col], errors='coerce').fillna(0)

print("✅ Conversion complete!")

# Also convert Opening Stock to numeric
df['Opening Stock'] = pd.to_numeric(df['Opening Stock'], errors='coerce').fillna(0)

print("\n📊 Basic statistics for key columns:")
print("="*50)
print("Opening Stock:")
print(df['Opening Stock'].describe())

print(f"\n📦 Sample of warehouse stock data:")
print(df[['Product Name'] + warehouse_cols[:5]].head(3))

🔍 Checking data types...
Product Name     object
msku             object
Opening Stock    object
Buffer Stock     object
TLCQ             object
BLR7             object
BLR8             object
BOM5             object
BOM7             object
CCU1             object
CCX1             object
DEL4             object
DEL5             object
DEX3             object
PNQ2             object
PNQ3             object
SDED             object
SDEE             object
XHJ9             object
dtype: object

🔢 Converting warehouse columns to numeric...
✅ Conversion complete!

📊 Basic statistics for key columns:
Opening Stock:
count    1011.000000
mean        6.597428
std        31.154631
min         0.000000
25%         0.000000
50%         0.000000
75%         2.000000
max       624.000000
Name: Opening Stock, dtype: float64

📦 Sample of warehouse stock data:
     Product Name  TLCQ  BLR7  BLR8  BOM5  BOM7
0  Chimmy Pillow    192     0     0     0    -1
1    Tata Pillow      0     0     0     0     0
2

### 📊 Quick Insights:

- 1,011 products in inventory
- Average stock: 6.6 units per product
- 75% of products have ≤2 units (very lean inventory!)
- Max stock: 624 units (one popular product!)
- ⚠️ Negative stock spotted: BOM7 shows -1 for Chimmy Pillow

Now let's analyze the warehouse stock distribution and find those problem areas:

In [7]:
# Warehouse stock analysis
warehouse_cols = ['TLCQ', 'BLR7', 'BLR8', 'BOM5', 'BOM7', 'CCU1', 'CCX1', 
                  'DEL4', 'DEL5', 'DEX3', 'PNQ2', 'PNQ3', 'SDED', 'SDEE', 'XHJ9']

print("🏭 WAREHOUSE STOCK ANALYSIS")
print("="*50)

# Total stock per warehouse
print("📦 Total stock per warehouse:")
warehouse_totals = df[warehouse_cols].sum().sort_values(ascending=False)
for warehouse, total in warehouse_totals.items():
    print(f"  {warehouse}: {total:>8.0f} units")

print(f"\n🎯 TOTAL INVENTORY ACROSS ALL WAREHOUSES: {warehouse_totals.sum():,.0f} units")

# Calculate total stock per product
df['total_stock'] = df[warehouse_cols].sum(axis=1)

print(f"\n📊 STOCK DISTRIBUTION:")
print(f"  Products with stock > 0: {len(df[df['total_stock'] > 0]):>4}")
print(f"  Products with zero stock: {len(df[df['total_stock'] == 0]):>4}")  
print(f"  Products with negative stock: {len(df[df['total_stock'] < 0]):>4}")

# Find negative stock issues
print(f"\n⚠️  NEGATIVE STOCK ALERTS:")
print("="*35)
for col in warehouse_cols:
    negative_items = df[df[col] < 0]
    if len(negative_items) > 0:
        print(f"\n🚨 {col}: {len(negative_items)} items with negative stock")
        for idx, row in negative_items.iterrows():
            print(f"    {row['Product Name']}: {row[col]}")

# Top 5 products by total stock
print(f"\n🏆 TOP 5 PRODUCTS BY TOTAL STOCK:")
print("="*40)
top_stock = df.nlargest(5, 'total_stock')[['Product Name', 'msku', 'total_stock']]
for idx, row in top_stock.iterrows():
    print(f"  {row['Product Name']}: {row['total_stock']:.0f} units")

# Products with zero stock everywhere
zero_stock = df[df['total_stock'] == 0]
print(f"\n📋 ZERO STOCK ANALYSIS:")
print(f"  {len(zero_stock)} products have no stock anywhere ({len(zero_stock)/len(df)*100:.1f}%)")
print(f"  Sample zero-stock products:")
for i, row in zero_stock.head(5).iterrows():
    print(f"    {row['Product Name']}")

🏭 WAREHOUSE STOCK ANALYSIS
📦 Total stock per warehouse:
  TLCQ:     5755 units
  BLR7:        0 units
  CCU1:        0 units
  PNQ3:        0 units
  SDED:        0 units
  DEX3:        0 units
  DEL5:        0 units
  SDEE:        0 units
  XHJ9:        0 units
  BOM5:       -1 units
  DEL4:       -2 units
  BOM7:       -4 units
  PNQ2:       -4 units
  CCX1:       -9 units
  BLR8:       -9 units

🎯 TOTAL INVENTORY ACROSS ALL WAREHOUSES: 5,726 units

📊 STOCK DISTRIBUTION:
  Products with stock > 0:  330
  Products with zero stock:  654
  Products with negative stock:   27

⚠️  NEGATIVE STOCK ALERTS:

🚨 TLCQ: 25 items with negative stock
    HBD Red MB: -1
    HP Brown MB: -22
    LaVieRose Black MB: -3
    Starwar Black MB: -1
    Spirated Black MB: -1
    RedGreen Butterfly SG: -3
    MH A5 Black MB: -2
    MH A8 Brown MB: -3
    Red wing SG: -1
    Butterfly red yellow SG : -2
    Strawberry Red: -1
    Dumbledore wand Musicbox: -2
    Lord Voldemort wand Musicbox: -1
    hermoine l

## 🏭 MAJOR FINDINGS:
### 📊 Inventory Concentration Crisis:

- 99.5% of stock is in TLCQ warehouse only (5,755 out of 5,726 units)
- All other warehouses are essentially empty or have negative stock
- Single point of failure - entire business depends on TLCQ

### ⚠️ SERIOUS Negative Stock Problem:

- 27 products oversold across 6 warehouses
- Worst case: HP Brown MB is -22 units (sold 22 more than available!)
- Data sync issues between sales and inventory systems

### 📋 Stock Portfolio Health:

- 64.7% of products (654/1011) have ZERO stock everywhere
- Only 32.6% (330/1011) products actually available to sell
- Business continuity risk - most catalog unavailable

### 🏆 Star Products:

- ThugLife Thin: 603 units (10.5% of total inventory!)
- HBD Brown MB: 380 units
- Harry Potter Round SG: 347 units

In [8]:
# MSKU Pattern Analysis
print("🏷️  MSKU PATTERN ANALYSIS")
print("="*50)

# Split MSKU into components
print("📋 Analyzing MSKU structure...")
df['msku_clean'] = df['msku'].astype(str)

# Split by underscore
msku_parts = df['msku_clean'].str.split('_', expand=True)
print(f"MSKU components found: {msku_parts.shape[1]} parts")

# Analyze each component
if msku_parts.shape[1] >= 4:
    df['prefix'] = msku_parts[0]  # Should be CSTE
    df['number'] = msku_parts[1]  # Sequential number
    df['category'] = msku_parts[2]  # Product category (MB, ST, etc.)
    df['description'] = msku_parts[3]  # Product description
    
    print(f"\n🔍 MSKU Component Analysis:")
    print(f"="*35)
    
    print(f"Prefix analysis:")
    print(df['prefix'].value_counts().head())
    
    print(f"\nCategory codes (what do they mean?):")
    category_counts = df['category'].value_counts()
    for cat, count in category_counts.items():
        print(f"  {cat}: {count:>3} products")
    
    print(f"\n📊 Category Distribution with Stock Status:")
    print("="*45)
    for cat in category_counts.index:
        cat_data = df[df['category'] == cat]
        has_stock = len(cat_data[cat_data['total_stock'] > 0])
        zero_stock = len(cat_data[cat_data['total_stock'] == 0])
        negative_stock = len(cat_data[cat_data['total_stock'] < 0])
        avg_stock = cat_data['total_stock'].mean()
        
        print(f"{cat:>4}: {len(cat_data):>3} total | "
              f"{has_stock:>3} stocked | {zero_stock:>3} zero | "
              f"{negative_stock:>2} negative | avg: {avg_stock:>5.1f}")

# Look at number patterns
print(f"\n🔢 MSKU Number Range:")
if 'number' in df.columns:
    # Convert to numeric for analysis
    df['number_numeric'] = pd.to_numeric(df['number'], errors='coerce')
    valid_numbers = df['number_numeric'].dropna()
    print(f"  Lowest: {valid_numbers.min():.0f}")
    print(f"  Highest: {valid_numbers.max():.0f}")
    print(f"  Range: {valid_numbers.max() - valid_numbers.min():.0f}")
    print(f"  Valid numbers: {len(valid_numbers)}/{len(df)}")

# Sample MSKUs by category
print(f"\n📋 Sample MSKUs by Category:")
print("="*35)
for cat in category_counts.head(5).index:
    sample = df[df['category'] == cat]['msku'].iloc[0]
    print(f"  {cat}: {sample}")

🏷️  MSKU PATTERN ANALYSIS
📋 Analyzing MSKU structure...
MSKU components found: 8 parts

🔍 MSKU Component Analysis:
Prefix analysis:
prefix
CSTE         1002
XEXY            6
Caaju           2
CSTE#0433       1
Name: count, dtype: int64

Category codes (what do they mean?):
  OT: 463 products
  ST: 175 products
  MB: 121 products
  SG: 109 products
  Wands:  38 products
  HD:  34 products
  Organizer:  26 products
  CLT:  15 products
  Org:   9 products
  Swaddle:   6 products
  Orgn:   6 products
  Accs:   6 products
  DS:   2 products
  Pokemon:   1 products

📊 Category Distribution with Stock Status:
  OT: 463 total | 131 stocked | 330 zero |  2 negative | avg:   2.4
  ST: 175 total |  81 stocked |  86 zero |  8 negative | avg:   6.3
  MB: 121 total |  33 stocked |  81 zero |  7 negative | avg:  10.0
  SG: 109 total |  24 stocked |  82 zero |  3 negative | avg:  10.6
Wands:  38 total |   5 stocked |  26 zero |  7 negative | avg:   5.6
  HD:  34 total |  15 stocked |  19 zero |  0 ne